pages/dev/general-message-passing/cosmos-gmp.mdx

```diff
@@ -0,0 +1,165 @@
+for _, receiver := range msg.ReceiverAddresses {
+    addresses = append(addresses, common.HexToAddress(receiver))
+}
+
+payload, err := abi.Arguments{{Type: addressesType}}.Pack(addresses)
+if err != nil {
+    return nil, err
+}
+
+message := Message{
+    DestinationChain:   msg.DestinationChain,
+    DestinationAddress: msg.DestinationAddress,
+    Payload:            payload,
+    Type:               TypeGeneralMessageWithToken,
+}
+
+bz, err := message.Marshal()
+if err != nil {
+    return nil, err
+}
+
+msg := ibctransfertypes.NewMsgTransfer(
+    ibctransfertypes.PortID,
+    "channel-17", // hard-coded channel id for demo
+    msg.Amount,
+    msg.Sender,
+    AxelarGMPAcc,
+    clienttypes.ZeroHeight(),
+    uint64(ctx.BlockTime().Add(6*time.Hour).UnixNano()),
+)
+msg.Memo = string(payload)
+
+res, err := k.ibcTransferK.Transfer(goCtx, msg)
+if err != nil {
+    return err
+}
+```
+
+## Messages from CosmWasm
+To send a message from CosmWasm Rust code, you must send a message to the Axelar Gateway (`axelar1dv4u5k73pzqrxlzujxg3qp8kvc3pje7jtdvu72npnt5zhq05ejcsn5qme5s`) via IBC.
+
+You should encode your payload using [ethabi](https://github.com/rust-ethereum/ethabi).
+
+```rust
+pub fn multi_send_to_evm(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    destination_chain: String,
+    destination_contract: String,
+    recipients: Vec<String>,
+    fee: Option<Fee>
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+
+    let addresses = recipients
+    .into_iter()
+    .map(|s| {
+        match s.parse::<H160>() {
+            Ok(address) => Ok(Token::Address(Address::from(address))),
+            Err(_) => Err(ContractError::InvalidRecipientAddress { address: s }),
+        }
+    })
+    .collect::<Result<Vec<Token>, ContractError>>()?;
+    let payload = encode(&[Token::Array(addresses)]);
+
+    let msg = GeneralMessage {
+        destination_chain,
+        destination_address: destination_contract.clone(),
+        payload,
+        type_: 2,
+        fee
+    };
+
+    let coin = cw_utils::one_coin(&info).unwrap();
+    let ibc_transfer = MsgTransfer {
+        source_port: "transfer".to_string(),
+        source_channel: config.channel.to_string(),
+        token: Some(coin.into()),
+        sender: env.contract.address.to_string(),
+        receiver: AXELAR_GATEWAY.to_string(),
+        timeout_height: None,
+        timeout_timestamp: Some(env.block.time.plus_seconds(604_800u64).nanos()),
+        memo: to_string(&msg).unwrap(),
+    };
+
+    // Base response
+    let response = Response::new()
+        .add_attribute("status", "ibc_message_created")
+        .add_attribute("ibc_message", format!("{:?}", ibc_transfer));
+
+    return Ok(response.add_message(ibc_transfer));
+}
+```
+
+## Chain Support
+For a Cosmos blockchain to support receiving GMP messages (naively or via CosmWasm), it should implement the appropriate middleware in Go as part of the consensus layer.
```

Comment:
native needs implement middleware,
cosmwasm only needs install the [general purpose middleware ](https://github.com/osmosis-labs/osmosis/tree/main/x/ibc-hooks)